h2 websockets backport for 2.4.x

icing opened the pull request:
Based on #364 

* `core`: add `ap_get_pollfd_from_conn()`, increment `mmn` number
* mod_http2:
  - add support for bootstrapping WebSockets via HTTP/2, as  described 
     in RFC 8441. 
  - new directive 'H2WebSockets on|off' has been  added. The feature is by default not enabled.
         As also discussed in the manual, this feature should work for setups
         using "ProxyPass backend-url upgrade=websocket" without further changes.
         Special server modules for WebSockets will have to be adapted,
         most likely, as the handling if IO events is different with HTTP/2.
  - HTTP/2 WebSockets are supported on platforms with native pipes. This
         excludes Windows.
  - small optimisation for output passing when sending DATA frames
  - new directive `H2ProxyRequests on|off` to enable handling
         of HTTP/2 requests in a forward proxy configuration.
         General forward proxying is enabled via `ProxyRequests`. If the
         HTTP/2 protocol is also enabled for such a server/host, this new
         directive is needed in addition.

* tests: add websocket tests, requires python `websockets` module version 10.4 or newer

icing committed:
Merge 1907696,1907697 from trunk:

  *) mod_http2: new directive 'H2MaxDataFrameLen n' to limit the maximum
     amount of response body bytes put into a single HTTP/2 DATA frame.
     Setting this to 0 places no limit (but the max size allowed by the
     protocol is observed).
     The module, by default, tries to use the maximum size possible, which is
     somewhat around 16KB. This sets the maximum. When less response data is
     available, smaller frames will be sent.

icing committed:
Merge r1909769 from trunk (plus sync mod_http2.xml)

  *) mod_http2: v2.0.15 with the following fixes and improvements
     - New directive 'H2EarlyHint name value' to add headers to a response,
       picked up already when a "103 Early Hints" response is sent. 'name' and
       'value' must comply to the HTTP field restrictions.
       This directive can be repeated several times and header fields of the
       same names add. Sending a 'Link' header with 'preload' relation will
       also cause a HTTP/2 PUSH if enabled and supported by the client.
     - Fixed an issue where requests were not logged and accounted in a timely
       fashion when the connection returns to "keepalive" handling, e.g. when
       the request served was the last outstanding one.
       This led to late appearance in access logs with wrong duration times
       reported.
     - Accurately report the bytes sent for a request in the '%O' Log format.
       This addresses #203, a long outstanding issue where mod_h2 has reported
       numbers over-eagerly from internal buffering and not what has actually
       been placed on the connection.
       The numbers are now the same with and without H2CopyFiles enabled.

icing committed:
Merge r1910157 from trunk

 * mod_proxy_http2: fixed using the wrong "bucket_alloc" from the backend
   connection when sending data on the frontend one. This caused crashes
   or infinite loops in rare situations.
 * mod_proxy_http2: fixed a bug in retry/response handling that could lead
   to wrong status codes or HTTP messages send at the end of response bodies
   exceeding the announced content-length.
 * mod_proxy_http2: fix retry handling to not leak temporary errors.
   On detecting that that an existing connection was shutdown by the other
   side, a 503 response leaked even though the request was retried on a
   fresh connection.
 * mod_http2: fixed a bug that did cleanup of consumed and pending buckets in
   the wrong order when a bucket_beam was destroyed.

icing committed:
Merge of 1910331,1910386,1910441 from trunk

  *) mod_http2: fixed a bug in flushing pending data on an already closed
     connection that could lead to a busy loop, preventing the HTTP/2 session
     to close down successfully. Fixed PR 66624.
     Fixed a bug that could lead to a crash in main connection
     output handling. This occured only when the last request on a HTTP/2
     connection had been processed and the session decided to shut down.
     This could lead to an attempt to send a final GOAWAY while the previous
     write was still in progress. See PR 66646.

icing committed:
*) mod_http2: added support for bootstrapping WebSockets via HTTP/2, as
     described in RFC 8441. A new directive 'H2WebSockets on|off' has been
     added. The feature is by default not enabled.
     As also discussed in the manual, this feature should work for setups
     using "ProxyPass backend-url upgrade=websocket" without further changes.
     Special server modules for WebSockets will have to be adapted,
     most likely, as the handling if IO events is different with HTTP/2.
     HTTP/2 WebSockets are supported on platforms with native pipes. This
     excludes Windows.

git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1910507 13f79535-47bb-0310-9956-ffa450edef68

icing committed:
*) mod_http2: adding checks for websocket support on platform and
     server versions. Give error message accordingly when trying to
     enable websockets in unsupported configurations.
     Add test and code to check the, finally selected, server of
     a request_rec for websocket support or 501 the request.



git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1910535 13f79535-47bb-0310-9956-ffa450edef68

icing committed:
mod_http2: optimization for c1 output passing when collecting
           next DATA frame for a response.



git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1910648 13f79535-47bb-0310-9956-ffa450edef68

icing committed:
HTTP/2 WebSockets: adding throughput tests and test client improvements



git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1910649 13f79535-47bb-0310-9956-ffa450edef68

icing committed:
tests: add the websockets python module version checks needed for our CI infra

git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1910654 13f79535-47bb-0310-9956-ffa450edef68

icing committed:
*) mod_http2: new directive `H2ProxyRequests on|off` to enable handling
     of HTTP/2 requests in a forward proxy configuration.
     General forward proxying is enabled via `ProxyRequests`. If the
     HTTP/2 protocol is also enabled for such a server/host, this new
     directive is needed in addition.



git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1910656 13f79535-47bb-0310-9956-ffa450edef68

icing committed:
lognos for the new debugs



git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1910658 13f79535-47bb-0310-9956-ffa450edef68

ylavic committed:
test_800_websockets: AH10467 is an expected error for the _fail_proto test.

So ignore it, and move the _fail_proto test first to avoid messing up with
the error logs of the other tests.



git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1910667 13f79535-47bb-0310-9956-ffa450edef68

icing committed:
mod_http2: move get_pollfd_from_conn hook outside the HAS_RESPONSE_BUCKET part


git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1910685 13f79535-47bb-0310-9956-ffa450edef68

icing committed:
adjust mmn checks to 2.4.x backport numbers

icing committed:
mod_http2: when failing a request, make sure the request_rec fields
are populated (best effort) for the response generation since this
might access fields in there.
Respect a http_status already assigned during stream processing
for error generation.



git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1910686 13f79535-47bb-0310-9956-ffa450edef68

icing committed:
mod_http2: add aplogno where makefile did not detect it missing?



git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1910688 13f79535-47bb-0310-9956-ffa450edef68

icing committed:
merge aplogno de-duplication fix from trunk

icing committed:
Sync test cases and their version checks with mod_h2

icing committed:
tests: sync with other branches

ylavic commented on modules/proxy/proxy_util.c:
```diff
@@ -4765,17 +4773,12 @@ PROXY_DECLARE(apr_status_t) ap_proxy_tunnel_create(proxy_tunnel_rec **ptunnel,
-    apr_socket_opt_set(tunnel->client->pfd->desc.s, APR_SO_NONBLOCK, 1);
-    apr_socket_opt_set(tunnel->origin->pfd->desc.s, APR_SO_NONBLOCK, 1);
+    tunnel->timeout = (origin_timeout >= 0 && origin_timeout > client_timeout)?
+                      origin_timeout : client_timeout;
```
I'm wondering if we shouldn't use the smaller of the two timeouts instead (i.e. change the original behaviour). It could be easier for the user to control the lifetime/ending of ws connections by doing this..

icing replied:
I was wondering about the original choice. I think it makes sense to use the smaller timeout, since this *should* tear down the tunnel, I believe.

ylavic replied:
Yeah, makes more sense to me.

icing replied:
Like this?

```
    tunnel->timeout = (client_timeout > 0 && client_timeout < origin_timeout)?
                      client_timeout : origin_timeout;
```

icing replied:
The question really is what a timeout of `0` actually should do? Would the tunnel not exit after the first poll without an event?

ylavic replied:
A zero timeout is probably not possible/useful but strictly speaking it would be:
```
    tunnel->timeout = (client_timeout >= 0 && client_timeout < origin_timeout ?
                       client_timeout : origin_timeout);
```
(using _my_ codestyle to please _my_ vim indent :)

ylavic replied:
> The question really is what a timeout of `0` actually should do? Would the tunnel not exit after the first poll without an event?

I think it would timeout immediately if data on any side are not available immediately (not very useful but an admin issue..).

icing replied:
Added as r1910704 to trunk.

ylavic replied:
:+1: thanks Stefan!

icing committed:
proxy: in proxy tunnels, use the smaller timeout value of
       client and origin as timeout for polling the tunnel.



git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1910704 13f79535-47bb-0310-9956-ffa450edef68

icing committed:
proxy_util: revert timeout selection on a proxy tunnel back to
take the large of client/origin values.

Add websocket test to verify that longer proxypass timeout is in
effect.



git-svn-id: https://svn.apache.org/repos/asf/httpd/httpd/trunk@1910809 13f79535-47bb-0310-9956-ffa450edef68

icing committed:
HTTP/2, websockets

- provide "Connection" header in internal Upgrade request
- set input notification if input is already available when
 stream processing is started
- add optional function in mod_http2.h to use instead of
 ap_get_pollfd_from_conn() in older servers with a new
 mod_http2

icing committed:
check new mmn version and allow websocket CONNECT, even without it

icing committed:
Latest sync with mod_h2, mostly cosmetic

rpluem commented on include/ap_mmn.h:
```diff
@@ -595,15 +596,15 @@
  * 20120211.126 (2.4.55-dev) Add additional hcmethod_t enums and PROXY_WORKER_IS_ERROR
  * 20120211.127 (2.4.56-dev) Add ap_proxy_canonenc_ex
- * 20120211.128 (2.4.55-dev) Add AP_CTIME_OPTION_GMTOFF to util_time.h
+ * 20120211.129 (2.4.58-dev) Add ap_get_pollfd_from_conn()
```
Why removing the 128 minor bump from comments?

icing replied:
Ah, my mistake. 128 was merged when the PR already existed and I messed this up during the rebase.

rpluem commented on modules/http2/h2_c2_filter.c:
```diff
@@ -108,20 +109,39 @@ apr_status_t h2_c2_filter_request_in(ap_filter_t *f,
+                      "h2_c2_filter_request_in(%s): adding request bucket",
+                      conn_ctx->id);
+        b = h2_request_create_bucket(req, f->r);
+        APR_BRIGADE_INSERT_TAIL(bb, b);
+
+        if (req->http_status != H2_HTTP_STATUS_UNSET) {
             /* error was encountered preparing this request */
-            b = ap_bucket_error_create(conn_ctx->request->http_status, NULL, f->r->pool,
+            ap_log_cerror(APLOG_MARK, APLOG_TRACE2, 0, f->c,
+                          "h2_c2_filter_request_in(%s): adding error bucket %d",
+                          conn_ctx->id, req->http_status);
+            b = ap_bucket_error_create(req->http_status, NULL, f->r->pool,
                                        f->c->bucket_alloc);
             APR_BRIGADE_INSERT_TAIL(bb, b);
             return APR_SUCCESS;
```
In the old code we only had the error bucket in `bb`in case that `req->http_status != H2_HTTP_STATUS_UNSET`. Now we always have a request bucket in `bb` even if `bb`in case that `req->http_status != H2_HTTP_STATUS_UNSET`. Does this matter?

icing replied:
This is the case with `AP_HAS_RESPONSE_BUCKETS`, so `trunk`. 

In trunk, we need to add a request bucket to the input, so the context for processing the error bucket can be established, e.g. if the error is a 400 response, what this response is for.

This case did not trigger in trunk before websockets and their tests. Now in `test_h2_800_01`, we send a `CONNECT` with an unknown `procotol:` header. That will be answered with a 400.

error buckets without an established request lead to a crash in trunk. I think we can discuss there how to improve the situation and if that could possibly apply to HTTP/1.1 processing as well.

rpluem replied:
Thanks for explaining

rpluem commented on modules/http2/h2_c2_filter.c:
```diff
@@ -184,7 +204,7 @@ static int uniq_field_values(void *d, const char *key, const char *val)
         for (i = 0, strpp = (char **) values->elts; i < values->nelts;
              ++i, ++strpp) {
-            if (*strpp && apr_strnatcasecmp(*strpp, start) == 0) {
+            if (*strpp && ap_cstr_casecmp(*strpp, start) == 0) {
```
This gives different results for strings e.g. that consist only of spaces, but off different length, or that start with a different amount of spaces as prefix and are otherwise the same.

icing replied:
Oh, then I misunderstood Yann's comment that `ap_cstr_casecmp` is the preferred use.

I will revert this instance in header merging. There *should* be no leading spaces, but...

rpluem commented on modules/http2/h2_c2_filter.c:
```diff
@@ -292,7 +312,7 @@ static h2_headers *create_response(request_rec *r)
         while (field && (token = ap_get_list_item(r->pool, &field)) != NULL) {
             for (i = 0; i < r->content_languages->nelts; ++i) {
-                if (!apr_strnatcasecmp(token, languages[i]))
+                if (!ap_cstr_casecmp(token, languages[i]))
```
Same as above

icing replied:
Will revert.

rpluem commented on modules/http2/h2_proxy_util.c:
```diff
@@ -381,7 +381,7 @@ static int iq_bubble_down(h2_proxy_iqueue *q, int i, int bottom,
  ******************************************************************************/
 #define H2_HD_MATCH_LIT_CS(l, name)  \
-    ((strlen(name) == sizeof(l) - 1) && !apr_strnatcasecmp(l, name))
+    ((strlen(name) == sizeof(l) - 1) && !ap_cstr_casecmp(l, name))
```
Same as above on this topic.

icing replied:
Wil revert.

rpluem commented on modules/http2/h2_proxy_util.c:
```diff
@@ -500,7 +500,7 @@ static int ignore_header(const literal *lits, size_t llen,
     for (i = 0; i < llen; ++i) {
         lit = &lits[i];
-        if (lit->len == nlen && !apr_strnatcasecmp(lit->name, name)) {
+        if (lit->len == nlen && !ap_cstr_casecmp(lit->name, name)) {
```
Same as above on this topic.

icing replied:
Will revert.

rpluem commented on modules/http2/h2_proxy_util.c:
```diff
@@ -542,7 +542,7 @@ void h2_proxy_util_camel_case_header(char *s, size_t len)
 /** Match a header value against a string constance, case insensitive */
 #define H2_HD_MATCH_LIT(l, name, nlen)  \
-    ((nlen == sizeof(l) - 1) && !apr_strnatcasecmp(l, name))
+    ((nlen == sizeof(l) - 1) && !ap_cstr_casecmp(l, name))
```
Same as above on this topic.

icing replied:
Will revert.

rpluem commented on modules/http2/h2_push.c:
```diff
@@ -426,7 +426,7 @@ static void inspect_link(link_ctx *ctx, const char *s, size_t slen)
 {
-    if (!apr_strnatcasecmp("link", key)) {
+    if (!ap_cstr_casecmp("link", key)) {
         inspect_link(ctx, value, strlen(value));
```
Same as above on this topic.

icing replied:
here it should be the better choice.

rpluem commented on modules/proxy/proxy_util.c:
```diff
@@ -4799,14 +4802,43 @@ PROXY_DECLARE(apr_status_t) ap_proxy_tunnel_create(proxy_tunnel_rec **ptunnel,
+        tunnel->client->pfd->reqevents = 0;
+        tunnel->origin->pfd->reqevents = APR_POLLIN | APR_POLLHUP |
+                                         APR_POLLOUT | APR_POLLERR;
+        if ((rv = apr_pollset_add(tunnel->pollset, tunnel->origin->pfd))) {
```
Don't we add the client fd aka the pipe to the pollset?

icing replied:
No, we cannot POLLOUT the pipe fd. It is ony way, only for input pending notifications.

rpluem commented on modules/http2/h2_ws.h:
```diff
@@ -0,0 +1,35 @@
+ * Rewrite a websocket request.
+ *
+ * @param req the h2 request to rewrite
+ * @param conn the connection to process the request on
```
Nitpick: Either `c2` here in the docs or `conn` below.

icing replied:
Will fix.

rpluem commented on modules/http2/h2_util.c:
```diff
@@ -1706,7 +1706,7 @@ static apr_status_t req_add_header(apr_table_t *headers, apr_pool_t *pool,
     }
     else if (nv->namelen == sizeof("cookie")-1
-             && !apr_strnatcasecmp("cookie", (const char *)nv->name)) {
+             && !ap_cstr_casecmp("cookie", (const char *)nv->name)) {
```
Same comment as above

icing replied:
here it is the better choice. We checked the length already.

rpluem commented on modules/http2/h2_util.c:
```diff
@@ -1725,7 +1725,7 @@ static apr_status_t req_add_header(apr_table_t *headers, apr_pool_t *pool,
     }
     else if (nv->namelen == sizeof("host")-1
-             && !apr_strnatcasecmp("host", (const char *)nv->name)) {
+             && !ap_cstr_casecmp("host", (const char *)nv->name)) {
```
Same comment as above

icing replied:
Better choice, length already checked.

rpluem commented on modules/http2/h2_util.c:
```diff
@@ -1650,7 +1650,7 @@ static int contains_name(const literal *lits, size_t llen, nghttp2_nv *nv)
         if (lit->len == nv->namelen
-            && !apr_strnatcasecmp(lit->name, (const char *)nv->name)) {
+            && !ap_cstr_casecmp(lit->name, (const char *)nv->name)) {
             return 1;
```
Same comment as above

icing replied:
Better choice. `nghttp2_nv` will not take field names with spaces.

rpluem commented on modules/http2/h2_util.h:
```diff
@@ -337,7 +337,7 @@ apr_size_t h2_util_table_bytes(apr_table_t *t, apr_size_t pair_extra);
 #define H2_HD_MATCH_LIT(l, name, nlen)  \
-    ((nlen == sizeof(l) - 1) && !apr_strnatcasecmp(l, name))
+    ((nlen == sizeof(l) - 1) && !ap_cstr_casecmp(l, name))
 
```
Same comment as above

icing replied:
Will revert for better-safe-than-sorry reasons since this macro is used in several places.

rpluem commented on modules/http2/h2_ws.c:
```diff
@@ -0,0 +1,361 @@
+                 * or in the request processings implementation of WebSockets */
+                ap_log_cerror(APLOG_MARK, APLOG_ERR, 0, c2, APLOGNO(10463)
+                              "h2_c2(%s-%d): websocket CONNECT, 101 response "
+                              "without 'Sec-WebSocket-Accept: %s' but expected %s",
```
```suggestion
                              "with 'Sec-WebSocket-Accept: %s' but expected %s",
```

icing replied:
Will fix.

covener commented on include/ap_mmn.h:
```diff
@@ -412,6 +412,7 @@
  *                         ap_condition_if_none_match(),
  *                         ap_condition_if_modified_since(),
  *                         ap_condition_if_range()
+<<<<<<< HEAD
```
leftover line from merge conflict

icing replied:
Thanks, will push fix.

icing committed:
Fixes after review by rpluem

- re-add ap_mmn comment for 128 minor version
- reverted most changes of apr_strnatcasecmp() to ap_cstr_casecmp()
  since the former has special whitespace handling which *may* mattter
  in certain cases
- fixed log message when wrong Sec-WebSocket-Accept response
  header was found

icing commented:
Many thanks for reviewing and finding mistakes. Just pushed the changes.

icing committed:
Remove line mistakenly left in by merge conflict.

icing committed:
increase curl buffering limit, as it changes for version used

icing committed:
Port more relaxed stutter check from trunk for test_h2_712

tititiou36 commented on modules/http2/h2_config.c:
```diff
@@ -363,6 +375,11 @@ static void h2_srv_config_seti(h2_config *conf, h2_config_var_t var, int val)
             break;
+        case H2_CONF_PROXY_REQUESTS:
+            H2_CONFIG_SET(conf, proxy_requests, val);
+        case H2_CONF_WEBSOCKETS:
```
Blind guess: break missing?

icing replied:
Nice catch. Will fix this silly mistake.

icing replied:
Pushed.

tititiou36 commented on modules/http2/h2_c2.c:
```diff
@@ -668,11 +718,21 @@ static apr_status_t c2_process(h2_conn_ctx_t *conn_ctx, conn_rec *c)
     apr_time_t timeout;
+    apr_status_t rv = APR_SUCCESS;
+
+    if(req->protocol && !strcmp("websocket", req->protocol)) {
```
Nit: Space after if

icing replied:
Thanks. Switch between projects with different coding styles poses challenges. ;)

icing replied:
Pushed.

icing committed:
Fixes after review by @tititiou36

rpluem-vf commented:
As this PR touches the HTTP/2 implementation I think it should pass the HTTP/2 test suite.

rpluem commented:
As this PR touches the HTTP/2 implementation I think it should pass the HTTP/2 test suite.

icing committed:
Increase delay in stutter check 712_02 for cpu starved CI runs

- eliminiate 712_03 as being then really the same as 712_02

icing commented:
> As this PR touches the HTTP/2 implementation I think it should pass the HTTP/2 test suite.

The HTTP/2 test suite should be more reliable, even when CPU resources are scarce. Pushed a fix.

icing commented:
github runner seem to have a Monday. Fixed the failure in test_h2_600_32 in trunk r1911964.

icing committed:
disable test_h2_600-32. only works reliable with r1911964 from trunk

icing committed:
more test stability adjustments for low cpu availabililty

icing commented:
Merged to 2.4.x in [r1911979](http://svn.apache.org/viewvc?view=revision&revision=1911979).
